### PR TITLE
Refacto of module permission checks

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | It appeared some module actions were displayed even if the employee was not allowed to use it.In this PR, we merge the permissions checks for the links to display and for the action execution. This means if a button is not displayed, I can be sure the action would not pass if called manually. This PR also updated the **upgrade** action, which is now related to "add".
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | Nope
| Deprecations? | Yes
| Fixed ticket? | /
| How to test?  | Play with the permissions related to the modules and checks the available actions on the module page are properly updated. ping @rGaillard

### CODE REVIEW

**Review by toutantic — COMMENTED** [review]
Works for me

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 1]
```diff
@@ -111,11 +126,54 @@ public function getCatalogModulesNames(array $filter = array())
         return array_keys($this->getCatalogModules($filter));
     }
 
+
+    /**
+     * Check the permissions of the current context (CLI or employee) for a module
+     *
+     * @param array $actions Actions to check
+     * @param string $name The module nale
```
Comment (eternoendless) [inline src/Adapter/Module/AdminModuleDataProvider.php]: nale? 😁

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 2]
```diff
@@ -48,12 +51,16 @@ class AdminModuleDataProvider implements ModuleInterface
 
     const _DAY_IN_SECONDS_ = 86400; /* Cache for One Day */
 
+    protected $module_actions = array('install', 'uninstall', 'enable', 'disable', 'enable_mobile', 'disable_mobile', 'reset', 'upgrade');
```
Comment (eternoendless) [inline src/Adapter/Module/AdminModuleDataProvider.php]: Please change `$module_actions` to `$moduleActions`

Comment (eternoendless) [reply src/Adapter/Module/AdminModuleDataProvider.php]: Please add a PhpDoc header on your new properties.

It is important that each property describes:
- A short summary of what it holds / its purpose
- Type

Documenting properties is important because
- It enables autocompletion by IDEs
- It eases code understanding

Example:
```php
   /**
     * Query string parameters ($_GET).
     *
     * @var \Symfony\Component\HttpFoundation\ParameterBag
     */
    public $query;
```

You can find out more about PhpDoc by reading the [proposed PSR-5 standard](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md).

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 3]
```diff
@@ -111,11 +126,54 @@ public function getCatalogModulesNames(array $filter = array())
...
+    }
+
+    /**
+     * Check the permissions of the current context (CLI or employee) for a specified action
+     *
+     * @param string $action The action called in the module
+     * @param string $name (Optionnal for 'install') The module name to check
+     * @return boolean
+     */
+    public function allowedAccess($action, $name = '')
```
Comment (eternoendless) [inline src/Adapter/Module/AdminModuleDataProvider.php]: I think method names are better when they start with a verb: `isAccessAllowed` or `isAllowed`

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 4]
```diff
@@ -111,11 +126,54 @@ public function getCatalogModulesNames(array $filter = array())
...
+     *
+     * @param array $actions Actions to check
+     * @param string $name The module nale
+     * @return array of allowed actions
+     */
+    protected function filterAllowedActions(array $actions, $name = '')
+    {
+        $allowedActions = array();
+        foreach (array_keys($actions) as $action_name) {
+            if ($this->allowedAccess($action_name, $name)) {
```
Comment (eternoendless) [inline src/Adapter/Module/AdminModuleDataProvider.php]: You're mixing camelCase and snake_case variables here. Please use camelCase 🙂

File: src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig [hunk 5]
```diff
@@ -31,47 +31,41 @@
...
       {{ 'Discover'|trans({}, 'Admin.Modules.Feature') }}
     </a>
   </div>
-  {% else %}
+  {% elseif urls|length %}
   <div class="btn-group form-action-button-container">
     <form class="btn-group form-action-button" method="post" action="{{ urls[url_active] }}">
       <button type="submit" class="btn btn-primary-reverse btn-primary-outline light-button module_action_menu_{{ url_active }}"
-          data-confirm_modal="module-modal-confirm-{{ name }}-{{ url_active }}" {% if (level < constant('PrestaShopBundle\\Security\\Voter\\PageVoter::LEVEL_UPDATE')) or (level < constant('PrestaShopBundle\\Security\\Voter\\PageVoter::LEVEL_DELETE') and (url_active not in ['configure', 'install', 'upgrade'])) %} disabled="disabled" {% endif %}>
+          data-confirm_modal="module-modal-confirm-{{ name }}-{{ url_active }}">
```
Comment (eternoendless) [inline src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig]: Why did you delete the level checks?

File: tests/Unit/Adapter/Module/AdminModuleDataProviderTest.php [hunk 6]
```diff
@@ -38,6 +38,7 @@ class AdminModuleDataProviderTest extends UnitTestCase
     private $addonsDataProviderS;
     private $categoriesProviderS;
     private $adminModuleDataProvider;
+    private $moduleDataProviderS;
```
Comment (eternoendless) [inline tests/Unit/Adapter/Module/AdminModuleDataProviderTest.php:41]: Why is there a capital S in the end? 🤔

**Review by eternoendless — CHANGES_REQUESTED** [review]

File: tests/Unit/Adapter/Module/AdminModuleDataProviderTest.php (hunk 6 above)
Comment (Quetzacoalt91) [reply tests/Unit/Adapter/Module/AdminModuleDataProviderTest.php:41]: S stands for stub.

File: src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig (hunk 5 above)
Comment (Quetzacoalt91) [reply src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig]: Because we should avoid logical checks in the templates.

The checks are now done in a single place, and we now only have the available actions in the templates. Before, we had to check in the template if we could display each button.

Comment (eternoendless) [reply src/PrestaShopBundle/Resources/views/Admin/Module/Includes/action_menu.html.twig]: Ok, I wanted to make sure of that, thanks 🙂

**Review by eternoendless — DISMISSED** [review]
Thanks!

**Review by LittleBigDev — APPROVED** [review]

**Comment by eternoendless** [comment]
Thank you @Quetzacoalt91
